### PR TITLE
TenantLoading in sequence, accept 422 RMB-522 RMB-523

### PR DESCRIPTION
**Pull request by adamdickmeiss**

### CODE REVIEW

File: domain-models-runtime/src/main/java/org/folio/rest/tools/utils/TenantLoading.java [hunk 1]
```diff
@@ -202,121 +202,144 @@ private static void endWithXHeaders(HttpClientRequest req, Map<String, String> h
...
+            return null;
+          }
+          break;
+        case RAW_PUT:
+        case RAW_POST:
+          break;
+      }
+      return id;
+    } catch (Exception ex) {
+      f.handle(Future.failedFuture("exception: " + ex.getLocalizedMessage()));
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/tools/utils/TenantLoading.java]: This swallows the stacktrace and does not add information to the message. My suggestion:
```suggestion
      f.handle(Future.failedFuture(ex));
```
Note: ex.getLocalizedMessage() uses the locale of the JVM, this is not the locale of the client; therefore ex.getMessage() should be used instead.

File: domain-models-runtime/src/main/java/org/folio/rest/tools/utils/TenantLoading.java [hunk 2]
```diff
@@ -202,121 +202,144 @@ private static void endWithXHeaders(HttpClientRequest req, Map<String, String> h
...
+      return null;
+    }
+  }
+
+  private static void handleException(Throwable ex, HttpMethod method, String uri, Future<Void> f) {
+    String diag = method.name() + " " + uri + ": " + ex.getMessage();
+    log.error(diag);
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/tools/utils/TenantLoading.java]: The stacktrace should be logged as well:
```suggestion
    log.error(diag, ex);
```

File: domain-models-runtime/src/test/java/org/folio/rest/tools/utils/TenantLoadingTest.java [hunk 3]
```diff
@@ -268,15 +268,36 @@ public void testPutFailure(TestContext context) {
 
     TenantLoading tl = new TenantLoading();
     tl.addJsonIdContent("loadRef", "tenant-load-ref", "data", "data");
-    putStatus = 422;
+    putStatus = 423;
```
Comment (julianladisch) [inline domain-models-runtime/src/test/java/org/folio/rest/tools/utils/TenantLoadingTest.java]: This number looks like a typo. Can you use 500 instead?

File: domain-models-runtime/src/test/java/org/folio/rest/tools/utils/TenantLoadingTest.java [hunk 4]
```diff
@@ -268,15 +268,36 @@ public void testPutFailure(TestContext context) {
...
+    TenantAttributes tenantAttributes = new TenantAttributes()
+      .withModuleTo("mod-1.0.0")
+      .withParameters(parameters);
+    Map<String, String> headers = new HashMap<String, String>();
+    headers.put("X-Okapi-Url-to", "http://localhost:" + Integer.toString(port));
+
+    TenantLoading tl = new TenantLoading();
+    tl.addJsonIdContent("loadRef", "tenant-load-ref", "data", "data");
+    putStatus = 423;
+    tl.withAcceptStatus(423);
```
Comment (julianladisch) [inline domain-models-runtime/src/test/java/org/folio/rest/tools/utils/TenantLoadingTest.java]: This number looks like a typo. Can you use 500 instead?

File: domain-models-runtime/src/main/java/org/folio/rest/tools/utils/TenantLoading.java [hunk 5]
```diff
@@ -202,121 +202,144 @@ private static void endWithXHeaders(HttpClientRequest req, Map<String, String> h
...
-      }
-      log.warn(method1.name() + " " + putUri.toString() + ": " + ex.getMessage());
-    });
-    endWithXHeaders(reqPut, headers, content);
+        resPut.exceptionHandler(ex -> handleException(ex, method1, putUri.toString(), f));
+      });
+      reqPut.exceptionHandler(ex -> handleException(ex, method1, putUri.toString(), f));
+      endWithXHeaders(reqPut, headers, content);
+    } catch (Exception ex) {
+      log.error(ex.getLocalizedMessage(), ex);
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/tools/utils/TenantLoading.java]:
```suggestion
      log.error(ex.getMessage(), ex);
```

File: domain-models-runtime/src/main/java/org/folio/rest/tools/utils/TenantLoading.java [hunk 6]
```diff
@@ -202,121 +202,144 @@ private static void endWithXHeaders(HttpClientRequest req, Map<String, String> h
...
+      f.handle(Future.failedFuture("exception: " + ex.getLocalizedMessage()));
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/tools/utils/TenantLoading.java]:
```suggestion
      f.handle(Future.failedFuture(ex));
```

File: domain-models-runtime/src/test/java/org/folio/rest/tools/utils/TenantLoadingTest.java [hunk 7]
```diff
@@ -268,15 +268,36 @@ public void testPutFailure(TestContext context) {
...
-  public void testOkStatusCode(TestContext context) {
+  public void testokWithAcceptStatus(TestContext context) {
+    Async async = context.async();
+    List<Parameter> parameters = new LinkedList<>();
+    parameters.add(new Parameter().withKey("loadRef").withValue("true"));
+    TenantAttributes tenantAttributes = new TenantAttributes()
+      .withModuleTo("mod-1.0.0")
+      .withParameters(parameters);
+    Map<String, String> headers = new HashMap<String, String>();
+    headers.put("X-Okapi-Url-to", "http://localhost:" + Integer.toString(port));
```
Comment (julianladisch) [inline domain-models-runtime/src/test/java/org/folio/rest/tools/utils/TenantLoadingTest.java]:
```suggestion
    headers.put("X-Okapi-Url-to", "http://localhost:" + port);
```

File: domain-models-runtime/src/test/java/org/folio/rest/tools/utils/TenantLoadingTest.java [hunk 8]
```diff
@@ -268,15 +268,36 @@ public void testPutFailure(TestContext context) {
...
+
+    TenantLoading tl = new TenantLoading();
+    tl.addJsonIdContent("loadRef", "tenant-load-ref", "data", "data");
+    putStatus = 423;
+    tl.withAcceptStatus(423);
+    tl.perform(tenantAttributes, headers, vertx, res -> {
+      context.assertTrue(res.succeeded());
+      async.complete();
+    });
```
Comment (julianladisch) [inline domain-models-runtime/src/test/java/org/folio/rest/tools/utils/TenantLoadingTest.java]: `Async async` is not needed:
tl.perform(tenantAttributes, headers, vertx, context.asyncAssertSuccess())

File: domain-models-runtime/src/test/java/org/folio/rest/tools/utils/TenantLoadingTest.java (hunk 7 above)
Comment (adamdickmeiss) [reply domain-models-runtime/src/test/java/org/folio/rest/tools/utils/TenantLoadingTest.java]: I believe this is resolved now.

Comment (julianladisch) [reply domain-models-runtime/src/test/java/org/folio/rest/tools/utils/TenantLoadingTest.java]: The unneccessary call to Integer.toString is still there, but it is a minor issue.

**Pull request merged**